### PR TITLE
server: support batch grpc messages.

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

This PR supports batch gRPC messages between TiKVs and TiDBs.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
Unit tests and integration tests.

## Does this PR affect documentation (docs) update? (mandatory)
No.

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/kvproto/pull/294
https://github.com/pingcap/tidb/pull/7821

## Benchmark result if necessary (optional)
default

### CODE REVIEW

**Comment by Connor1996** [comment]
Please fix the conflicts

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -62,17 +77,27 @@ impl<T: RaftStoreRouter + 'static, E: Engine> Service<T, E> {
...
         ch: T,
         snap_scheduler: Scheduler<SnapTask>,
+        mut heavy_load_threshold: usize,
+        collect_runtime: Arc<Runtime>,
+        thread_load: Arc<ThreadLoad>,
     ) -> Self {
+        // Because here we returns messages to clients, `heavy_load_threshold` shouldn't be too
+        // small. We set the lower limit to 100 for now.
+        // TODO: make it configuable and fix the best default value.
+        heavy_load_threshold = ::std::cmp::max(heavy_load_threshold, 100);
```
Comment (siddontang) [inline src/server/service/kv.rs]: use a const value for 100

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -1138,6 +767,996 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+                id,
+                resp,
+                tx,
+                timer,
+                thread_load,
+                heavy_load_threshold,
+            );
+        }
+        BatchCommandsRequest_Request_oneof_cmd::Import(_) => {
+            panic!("unimplemented");
```
Comment (Connor1996) [inline src/server/service/kv.rs]: should it be implemented now?

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -1138,6 +753,784 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+                    id,
+                    req,
+                    tx.clone(),
+                    Arc::clone(&thread_load),
+                );
+            }
+            future::ok::<_, _>(())
+        });
+
+        ctx.spawn(request_handler.map_err(|e| error!("batch_commands error: {}", e)));
```
Comment (rleungx) [inline src/server/service/kv.rs]:
```suggestion
        ctx.spawn(request_handler.map_err(|e| error!("batch commands error: {}", e)));
```

File: src/server/service/kv.rs [hunk 4]
```diff
@@ -1138,6 +753,784 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+                .map_err(|e| debug!("{} failed: {:?}", "coprocessor_stream", e)),
+        );
+    }
+}
+
+fn response_batch_commands_request<F>(
+    executor: &TaskExecutor,
+    id: u64,
+    resp: F,
+    tx: Sender<(u64, BatchCommandsResponse_Response)>,
```
Comment (rleungx) [inline src/server/service/kv.rs]: `BatchCommandsResponse_Response` looks wired

File: src/server/service/kv.rs [hunk 5]
```diff
@@ -1138,6 +753,784 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+        res.cmd = Some(resp);
+        if tx.send((id, res)).is_err() {
+            error!("KvService response batch commands fail");
+            return Err(());
+        }
+        timer.observe_duration();
+        if let Some(notifier) = tx.get_notifier() {
+            if thread_load.in_heavy_load() {
+                executor1.spawn(
+                    Delay::new(Instant::now() + Duration::from_millis(1))
```
Comment (rleungx) [inline src/server/service/kv.rs]: better to use a constant?

File: src/server/metrics.rs [hunk 6]
```diff
@@ -75,6 +75,16 @@ lazy_static! {
         "Total number of handle grpc message failure",
         &["type"]
     ).unwrap();
+    pub static ref GRPC_REQ_BATCH_COMMANDS_SIZE: Histogram = register_histogram!(
+        "tikv_server_grpc_req_super_batch_size",
+        "Super batch size of gRPC requests",
```
Comment (zhangjinpeng1987) [inline src/server/metrics.rs]: Please remove the word `Super`.

File: src/server/service/kv.rs (hunk 2 above)
Comment (siddontang) [reply src/server/service/kv.rs]: I think we don't need to support it now. 
Btw, use `unimplemented!()` instead

**Comment by siddontang** [comment]
can you split this into two PRs? one is for refactoring futures, one is for batch.

File: src/server/service/kv.rs (hunk 4 above)
Comment (hicqu) [reply src/server/service/kv.rs]: I agree but it's generated by protobuf. :(

File: src/server/metrics.rs [hunk 7]
```diff
@@ -75,6 +75,16 @@ lazy_static! {
         "Total number of handle grpc message failure",
         &["type"]
     ).unwrap();
+    pub static ref GRPC_REQ_BATCH_COMMANDS_SIZE: Histogram = register_histogram!(
+        "tikv_server_grpc_req_super_batch_size",
```
Comment (zhangjinpeng1987) [inline src/server/metrics.rs]: Please remove the word `super`.

File: src/server/metrics.rs [hunk 8]
```diff
@@ -75,6 +75,16 @@ lazy_static! {
...
+        "grpc batch size of gRPC requests",
+        exponential_buckets(1f64, 2f64, 10).unwrap()
+    ).unwrap();
+    pub static ref GRPC_RESP_BATCH_COMMANDS_SIZE: Histogram = register_histogram!(
+        "tikv_server_grpc_resp_super_batch_size",
```
Comment (zhangjinpeng1987) [inline src/server/metrics.rs]: Ditto.

**Comment by zhangjinpeng1987** [comment]
rest LGTM
Please provide the benchmark result.

File: src/server/service/kv.rs [hunk 9]
```diff
@@ -727,6 +747,274 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+        if tx.send((id, res)).is_err() {
+            error!("KvService response batch commands fail");
+            return Err(());
+        }
+        timer.observe_duration();
+        if let Some(notifier) = tx.get_notifier() {
+            if thread_load.in_heavy_load() {
+                executor1.spawn(
+                    Delay::new(Instant::now() + DELAY_DURATION)
+                        .map_err(|_| error!("BatchCommands RPC delay responses error"))
```
Comment (MyonKeminta) [inline src/server/service/kv.rs]: Can we provide more information in the error logs? For example, print the `Error` out

**Pull request merged**